### PR TITLE
blas/gonum: add Dznrm2, Dzasum, Izamax, Zdotu

### PR DESCRIPTION
**Pull request by vladimir-ch**

### CODE REVIEW

**Comment by btracey** [comment]
I don't have time to look at this immediately, but aren't level 1 BLAS functions so much more fun than Lapack functions?

**Comment by vladimir-ch** [comment]
I enjoyed working on Dgeev, it was (almost always) fun for me. I like Lapack. Complex level 1 BLAS is kind of mechanical, slightly annoying due to complex numbers. I'll see the what about 2&3.

**Comment by btracey** [comment]
I liked doing `Dpotrf`, and `Dgelqf` and co were fun. `Dgesvd` itself was not at all fun, and `Dlasq*` et al were the worst. I'm working on the banded version of `Dpotrf` and friends, we'll see how much fun that is.

File: blas/gonum/level1cmplx128.go [hunk 1]
```diff
@@ -5,9 +5,163 @@
...
+				max = absV
+				idx = i
+			}
+		}
+		return idx
+	}
+	ix := incX
+	for i := 1; i < n; i++ {
+		v := x[ix]
+		absV := math.Abs(real(v)) + math.Abs(imag(v))
```
Comment (kortschak) [inline blas/gonum/level1cmplx128.go]: `dcabs1`?

File: blas/gonum/level1cmplx128.go [hunk 2]
```diff
@@ -5,9 +5,163 @@
...
+	}
+	if len(x) <= (n-1)*incX {
+		panic(badX)
+	}
+	idx := 0
+	max := math.Abs(real(x[0])) + math.Abs(imag(x[0]))
+	if incX == 1 {
+		for i := 1; i < n; i++ {
+			v := x[i]
+			absV := math.Abs(real(v)) + math.Abs(imag(v))
```
Comment (kortschak) [inline blas/gonum/level1cmplx128.go]: `dcabs1`?

File: blas/gonum/level1cmplx128.go [hunk 3]
```diff
@@ -124,3 +278,44 @@ func (Implementation) Zdotc(n int, x []complex128, incX int, y []complex128, inc
 	}
 	return c128.DotcInc(x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
 }
+
+// Zdotu computes the dot product
+//  x^T · y
+// of two complex vectors x and y.
+func (Implementation) Zdotu(n int, x []complex128, incX int, y []complex128, incY int) (dotu complex128) {
```
Comment (kortschak) [inline blas/gonum/level1cmplx128.go]: Does this need a named return?

File: blas/testblas/dzasum.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+			want := float64(n * (2*n + 1))
+			got := impl.Dzasum(n, x, incX)
+
+			if incX < 0 {
+				if got != 0 {
+					t.Errorf("Case n=%v,incX=%v: non-zero result when incX < 0. got %v", n, incX, got)
+				}
+				continue
+			}
+			if math.Abs(got-want) > 1e-14 {
```
Comment (kortschak) [inline blas/testblas/dzasum.go]: We use floats functions elsewhere, is there a reason not to?

Can you make the tol a const at the head of the function please.

File: blas/testblas/dznrm2.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	} {
+		n := len(test.x)
+		for _, incX := range []int{-10, -1, 1, 2, 9, 17} {
+			aincX := abs(incX)
+			var x []complex128
+			if n > 0 {
+				x = make([]complex128, (n-1)*aincX+1)
+			}
+			for i := range x {
+				x[i] = complex(math.NaN(), math.NaN())
```
Comment (kortschak) [inline blas/testblas/dznrm2.go]: `x[i] = cmplx.NaN()`

File: blas/testblas/dzasum.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	rnd := rand.New(rand.NewSource(1))
+	for _, n := range []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 50, 100} {
+		for _, incX := range []int{-5, 1, 2, 10} {
+			aincX := abs(incX)
+			var x []complex128
+			if n > 0 {
+				x = make([]complex128, (n-1)*aincX+1)
+			}
+			for i := range x {
+				x[i] = complex(math.NaN(), math.NaN())
```
Comment (kortschak) [inline blas/testblas/dzasum.go]: `x[i] = cmplx.NaN()`

File: blas/testblas/dznrm2.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+			}
+			im := make([]float64, n)
+			for i := range im {
+				im[i] = rnd.NormFloat64()
+			}
+			want := math.Hypot(impl.Dnrm2(n, re, 1), impl.Dnrm2(n, im, 1))
+
+			x := make([]complex128, (n-1)*incX+1)
+			for i := range x {
+				x[i] = complex(math.NaN(), math.NaN())
```
Comment (kortschak) [inline blas/testblas/dznrm2.go]: `x[i] = cmplx.NaN()`

File: blas/testblas/izamax.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	rnd := rand.New(rand.NewSource(1))
+	for _, n := range []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 50, 100} {
+		for _, incX := range []int{-5, 1, 2, 10} {
+			aincX := abs(incX)
+			var x []complex128
+			if n > 0 {
+				x = make([]complex128, (n-1)*aincX+1)
+			}
+			for i := range x {
+				x[i] = complex(math.NaN(), math.NaN())
```
Comment (kortschak) [inline blas/testblas/izamax.go]: `x[i] = cmplx.NaN()`

File: blas/testblas/zdotu.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+			}
+			xCopy := make([]complex128, len(x))
+			copy(xCopy, x)
+
+			var y []complex128
+			if n > 0 {
+				y = make([]complex128, (n-1)*aincY+1)
+			}
+			for i := range y {
+				y[i] = complex(math.NaN(), math.NaN())
```
Comment (kortschak) [inline blas/testblas/zdotu.go]: `x[i] = cmplx.NaN()`

File: blas/gonum/level1cmplx128.go (hunk 3 above)
Comment (vladimir-ch) [inline blas/gonum/level1cmplx128.go]: Copy/paste from `cmplx.go`. Done.

File: blas/gonum/level1cmplx128.go [hunk 10]
```diff
@@ -5,9 +5,163 @@
...
+		}
+		return 0
+	}
+	var sum float64
+	if incX == 1 {
+		if len(x) < n {
+			panic(badX)
+		}
+		for _, v := range x[:n] {
+			sum += math.Abs(real(v)) + math.Abs(imag(v))
```
Comment (vladimir-ch) [inline blas/gonum/level1cmplx128.go]: Done all.

File: blas/gonum/level1cmplx128.go [hunk 11]
```diff
@@ -5,9 +5,163 @@
...
+		}
+		panic(negativeN)
+	}
+	if len(x) <= (n-1)*incX {
+		panic(badX)
+	}
+	idx := 0
+	max := math.Abs(real(x[0])) + math.Abs(imag(x[0]))
+	if incX == 1 {
+		for i := 1; i < n; i++ {
```
Comment (vladimir-ch) [inline blas/gonum/level1cmplx128.go]: Done.

File: blas/testblas/dzasum.go (hunk 6 above)
Comment (vladimir-ch) [inline blas/testblas/dzasum.go]: I should probably check what `cmplx` has. Done.

File: blas/testblas/dzasum.go (hunk 4 above)
Comment (vladimir-ch) [inline blas/testblas/dzasum.go]: Done.

**Comment by vladimir-ch** [comment]
PTAL

**Pull request merged**